keps/sig-node/0000-apparmor-ga/README.md

```diff
@@ -0,0 +1,685 @@
+
+AppArmor is not available on every Linux distribution. Beside this, container
+runtimes have AppArmor as compile-time feature which may be disabled as well.
+With the GA API we do not change the error handling and behave exactly the same
+as the current error propagation paths.
+
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
+includes the Pod API, which specifies what profile the containers run with, and
+the `PodSecurityPolicy` API which specifies allowed profiles & a default
+profile. There are no changes planned to the current `PodSecurityPolicy` API
+as part of this KEP.
+
+#### Pod API
+
+The Pod AppArmor API is generally immutable, except in `PodTemplates`.
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The AppArmor options to use by this container. If AppArmor options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // +optional
+    AppArmor  *AppArmorProfile
+    ...
+}
+
+// Only one profile source may be set.
+// +union
+type AppArmorProfile struct {
+    // +unionDescriminator
+    Type AppArmorProfileType
+
+    // Load a profile defined on the node.
+    // The profile must be available on the node to work.
+    // The length of the profile is limited to 253 characters.
+    // The `Type` in this struct has to be set to `AppArmorProfileLocalhost`.
+    // +optional
+    LocalhostProfile *string
+}
+
+type AppArmorProfileType string
+
+const (
+    AppArmorProfileTypeUnconfined     AppArmorProfileType = "Unconfined"
+    AppArmorProfileTypeRuntimeDefault AppArmorProfileType = "RuntimeDefault"
+    AppArmorProfileTypeLocalhost      AppArmorProfileType = "Localhost"
+)
+```
+
+This API makes the options more explicit and leaves room for new profile sources
+to be added in the future (e.g. Kubernetes predefined profiles or ConfigMap
+profiles) and for future extensions, such as defining the behavior when a
+profile cannot be set.
+
+##### Localhost Profile
+
+This KEP proposes we GA LocalhostProfile as the only source of user-defined
+profiles at this point. User-defined profiles are essential for users to realize
+the full benefits out of AppArmor, allowing them to decrease their attack
+surface based on their own workloads.
+
+###### Updating AppArmor profiles
+
+AppArmor profiles are applied at container creation time. Therefore, updating
+the file contents of a profile on disk after that point will not cause any
+changes to the running containers that are using it, but will apply the updated
+profile on container restart. Note that amending such files is not recommended
+and may cause containers to fail on next restart, in the case of the new profile
+being more restrictive, invalid or the file no longer being present on disk.
```

Comment:
AppArmor works differently from seccomp, and isn't file-based. Although the profiles are defined in files, to be used they must be loaded into the kernel (e.g. using the `apparmor-loader` program).

I'm actually not sure what happens if you reload the profile that's in use.

Reply:
Updated the paragraph to reflect how profiles are loaded.

Reply:
Did some research about the replacement and the AppArmor FAQ seems to answer this question:
https://gitlab.com/apparmor/apparmor/-/wikis/FAQ#what-is-profile-replacement

>  ### What is Profile Replacement?
>
> Profile replacement allows for replacing a currently loaded profile with a new version of the same profile. When this is done all tasks using the old profile begin using the new profile for all access decisions going forward. Doing this has several security implications that the administrator must be aware of:
> * Tasks retain access to all resources they had open under the older version of the profile
> * Any children tasks (applications a task has launched) retain the confinement that was decided by the old version of the profile. For children inheriting the parent profile (ix), then new children tasks will use the new profile.
> 
> The most secure solution is to always restart services after a profile has been updated, but in practice this is not always desired or required, which is why replacement is provided despite the above limitations